cli/azd/pkg/lazy/lazy.go

```diff
@@ -0,0 +1,52 @@
+package lazy
+
+import "sync"
+
+type InitializerFn[T comparable] func() (T, error)
+
+// A data structure that will lazily load an instance of the underlying type
+// from the specified initializer
+type Lazy[T comparable] struct {
+	initialized bool
+	initializer InitializerFn[T]
+	value       T
+	error       error
+	mutex       sync.Mutex
+}
+
+// Creates a new Layz[T]
+func NewLazy[T comparable](initializerFn InitializerFn[T]) *Lazy[T] {
+	return &Lazy[T]{
+		initializer: initializerFn,
+	}
+}
+
+// Gets the value of the configured initializer
+// Initializer will only run once on success
+func (l *Lazy[T]) GetValue() (T, error) {
+	// Only allow a single caller to get a value at one time.
+	// Additional calls will block until current call is complete
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	if !l.initialized {
+		value, err := l.initializer()
+		if err == nil {
+			l.value = value
+			l.error = nil
+			l.initialized = true
+		} else {
+			l.error = err
+			l.initialized = false
+		}
+	}
+
+	return l.value, l.error
+}
+
+// Sets a value on the lazy type
+func (l *Lazy[T]) SetValue(value T) {
+	l.value = value
+	l.error = nil
+	l.initialized = true
```

Comment:
I think we also need to make SetValue() a critical path with a mutex.lock?

We must avoid a race when calling GetValue and SetValue at the same time right?

Reply:
I was thinking the same, but `SetValue()` is literally just setting a value assignment and doesn't have any other side effects and shouldn't be needed.

Reply:
@vhvb1989 Have you looked at simply using `sync.Once` which has proper multithreading support + guarantees at least-once initialization (initializers block when initialization is in-progress).

My personal preference is to not introduce `Lazy`. Instead, the owning object would use `sync.Once` + `val` to ensure one-time initialization. This allows flexibility of different initializations. If we did introduce `Lazy`, it would be a wrapper around `sync.Once` + `val`.

Reply:
From my research on `sync.Once` it ensures that the function declared in `sync.Once(func...)` is only ever run one time.  If we do decide to take this `Lazy[T]` that isn't exactly the goal.  The initializer could run more than one time, we just want to prevent multiple calls concurrently which the `mutex` solves for at this time allowing the first call to complete, and the 2nd call would simply return the internal cached value assuming a previous call had completed successfully.

Reply:
I think there are valid use cases for `Lazy[T]` data structure, just not 100% sold on the `Credentials` use case.

Reply:
> I was thinking the same, but `SetValue()` is literally just setting a value assignment and doesn't have any other side effects and shouldn't be needed.

The problem is calling SetValue() from two places at the same time.  Or calling SetValue and GetValue at the same time.  

The `T` should be atomic for this to be reliable as a general-purpose component

Reply:
Added the mutex for the `SetValue` operation.